hardware_interface/mainpage.dox

```diff
@@ -0,0 +1,303 @@
+
+![hw_interface_2.png](https://raw.githubusercontent.com/wiki/ros-controls/ros_control/hw_interface_2.png)
+
+As the image above suggests, a custom robot hardware interface is not
+limited to be composed of only one single interface. The robot can provide as
+many interfaces as required. It could for example provide both the
+\ref hardware_interface::PositionJointInterface and the
+\ref hardware_interface::VelocityJointInterface, and many more.
+
+Assuming a robot with two joints: A & B where both joints are position controlled.
+Such a robot is defined with a class derived from \ref hardware_interface::RobotHW
+that should provide the standard [PositionJointInterface](\ref hardware_interface::PositionJointInterface) and the
+[JointStateInterface](\ref hardware_interface::JointStateInterface), so it can re-use all controllers that are already
+written to work with the [PositionJointInterface](\ref hardware_interface::PositionJointInterface) and the
+[JointStateInterface](\ref hardware_interface::JointStateInterface). The code would look like this:
+
+\code
+#include <hardware_interface/joint_command_interface.h>
+#include <hardware_interface/joint_state_interface.h>
+#include <hardware_interface/robot_hw.h>
+
+class MyRobot : public hardware_interface::RobotHW
+{
+public:
+  MyRobot() 
+  { 
+    // Initialization of the robot's resources (joints, sensors, actuators) and
+    // interfaces can be done here or inside init().
+  }
+
+  bool init(ros::NodeHandle &root_nh, ros::NodeHandle &robot_hw_nh)
+  {
+    // Create a JointStateHandle for each joint and register them with the 
+    // JointStateInterface.
+    hardware_interface::JointStateHandle state_handle_a("A", &pos[0], &vel[0], &eff[0]);
+    jnt_state_interface.registerHandle(state_handle_a);
+
+    hardware_interface::JointStateHandle state_handle_b("B", &pos[1], &vel[1], &eff[1]);
+    jnt_state_interface.registerHandle(state_handle_b);
+
+    // Register the JointStateInterface containing the read only joints
+    // with this robot's hardware_interface::RobotHW.
+    registerInterface(&jnt_state_interface);
+
+    // Create a JointHandle (read and write) for each controllable joint
+    // using the read-only joint handles within the JointStateInterface and 
+    // register them with the JointPositionInterface.
+    hardware_interface::JointHandle pos_handle_a(jnt_state_interface.getHandle("A"), &cmd[0]);
+    jnt_pos_interface.registerHandle(pos_handle_a);
+
+    hardware_interface::JointHandle pos_handle_b(jnt_state_interface.getHandle("B"), &cmd[1]);
+    jnt_pos_interface.registerHandle(pos_handle_b);
+
+    // Register the JointPositionInterface containing the read/write joints
+    // with this robot's hardware_interface::RobotHW.
+    registerInterface(&jnt_pos_interface);
+
+    return true;
+  }
+
+private:
+  hardware_interface::JointStateInterface jnt_state_interface;
+  hardware_interface::PositionJointInterface jnt_pos_interface;
```

Comment:
Why exactly are both needed? Isn't [PositionJointInerface](http://docs.ros.org/melodic/api/hardware_interface/html/c++/classhardware__interface_1_1PositionJointInterface.html) sufficent because it inherits from [JointCommandInterface](http://docs.ros.org/melodic/api/hardware_interface/html/c++/classhardware__interface_1_1JointCommandInterface.html) which can be used for reading and writing? Or are both required because all [existing controllers](https://github.com/ros-controls/ros_controllers) require a `JointStateInterface` to read the state of the robot's resources and don't make use of read/write interfaces (only custom controllers could)?

Reply:
`JointStateInterface` gives you read access to all joint values without conflicting with other controllers. If you were to use a PositionJointInterface but only ever intended it for reading positions from it, other controllers can't write to it either as you reserved it for writing with this interface.